File: elixir/rpc_client.exs
defmodule FibonacciRpcClient do
  def wait_for_messages(_channel, correlation_id) do
    receive do
      {:basic_deliver, payload, %{correlation_id: ^correlation_id}} ->
        {n, _} = Integer.parse(payload)
        n
    end
  end
  def call(n) do
    {:ok, connection} = AMQP.Connection.open
    {:ok, channel} = AMQP.Channel.open(connection)

    {:ok, %{queue: queue_name}} = AMQP.Queue.declare(channel, "", exclusive: true)
    AMQP.Basic.consume(channel, queue_name, nil, no_ack: true)
    correlation_id = :erlang.unique_integer |> :erlang.integer_to_binary |> Base.encode64
    request = to_string(n)
    AMQP.Basic.publish(channel, "", "rpc_queue", request, reply_to: queue_name, correlation_id: correlation_id)

    FibonacciRpcClient.wait_for_messages(channel, correlation_id)
  end
end

num =
  case System.argv do
    []    -> 30
    param ->
      {x, _} =
        param
        |> Enum.join(" ")
        |> Integer.parse
      x
  end

IO.puts " [x] Requesting fib(#{num})"
response = FibonacciRpcClient.call(num)
IO.puts " [.] Got #{response}"

File: erlang/src/rpc_client.erl
-module(rpc_client).
-export([start/1]).

-include_lib("amqp_client/include/amqp_client.hrl").

start(Argv) ->
    Num = case Argv of
             [] -> 10;
             [Arg] -> list_to_integer(Arg)
          end,
    io:format(" [x] Requesting fib(~p)~n", [Num]),
    Response = call(Num),
    io:format(" [.] Got ~p~n", [Response]),
    ok.

call(Num) ->
    {ok, Connection} =
        amqp_connection:start(#amqp_params_network{host = "localhost"}),
    {ok, Channel} = amqp_connection:open_channel(Connection),
    RequestQueue = <<"rpc_queue">>,
    CorrelationId = uuid:get_v4(),

    amqp_channel:call(Channel, #'queue.declare'{queue = RequestQueue}),

    #'queue.declare_ok'{queue = ReplyQueue} =
        amqp_channel:call(Channel, #'queue.declare'{exclusive = true}),

    Method = #'basic.consume'{queue = ReplyQueue, no_ack = true},
    amqp_channel:subscribe(Channel, Method, self()),

    amqp_channel:cast(Channel,
                      #'basic.publish'{
                        exchange = <<>>,
                        routing_key = RequestQueue},
                      #amqp_msg{
                        payload = integer_to_binary(Num),
                        props = #'P_basic'{
                            reply_to = ReplyQueue,
                            correlation_id = CorrelationId}
                        }),
    
    Response = wait_for_messages(CorrelationId),

    amqp_channel:close(Channel),
    amqp_connection:close(Connection),

    Response.

wait_for_messages(CorrelationId) ->
    receive
        {#'basic.deliver'{}, #amqp_msg{payload = Body, props = Props}} ->
            #'P_basic'{correlation_id = CorrId} = Props,
            if CorrelationId == CorrId ->
                binary_to_integer(Body);
            true ->
                -1
            end
    end.
